sdk/tracing/src/main/java/io/opentelemetry/sdk/trace/export/BatchSpanProcessor.java

```diff
@@ -276,34 +287,34 @@ private void exportBatches(ArrayList<ReadableSpan> spanList) {
     }
 
     // Exports the list of SpanData to the SpanExporter.
-    @SuppressWarnings("BooleanParameter")
     private void onBatchExport(final List<SpanData> spans) {
-      if (exportAvailable.compareAndSet(true, false)) {
-        try {
-          final CompletableResultCode result = spanExporter.export(spans);
-          result.whenComplete(
-              new Runnable() {
-                @Override
-                public void run() {
-                  if (!result.isSuccess()) {
-                    logger.log(Level.FINE, "Exporter failed");
-                  }
-                  exportAvailable.set(true);
-                }
-              });
-          timer.schedule(
-              new TimerTask() {
-                @Override
-                public void run() {
-                  result.fail();
-                }
-              },
-              exporterTimeoutMillis);
-        } catch (Exception e) {
-          logger.log(Level.WARNING, "Exporter threw an Exception", e);
+      try {
+        final CompletableResultCode result = spanExporter.export(spans);
+        final CountDownLatch latch = new CountDownLatch(1);
+        result.whenComplete(
+            new Runnable() {
+              @Override
+              public void run() {
+                latch.countDown();
+              }
+            });
+        timer.schedule(
+            new TimerTask() {
+              @Override
+              public void run() {
+                result.fail();
+              }
+            },
+            exporterTimeoutMillis);
+        latch.await(exporterTimeoutMillis, TimeUnit.MILLISECONDS);
```

Comment:
In essence, we're re-introducing blocking behaviour. I see a better way, but need think a bit more...

Reply:
Yeah, the way I think about it is BSP is itself an asynchronous span processor since it uses a background thread to process a queue. It has two main properties, a queue size, and a export batch size. If the queue is full, spans are dropped. If the backend cannot support large batches, the export batch size can be made smaller than the queue size. This is to prevent too much pressure on the backend, so blocking is probably the right behavior (sending all the mini batches concurrently is presumably at least as bad as the large batch size the setting tries to avoid), we literally want to slow down the export for the backend. And if it's too slow, then the queue will fill up and spans will be dropped - but that's the balancing act of the backend performance vs ingestion rate. Because BSP uses a background thread, it's impossible for this blocking to leak into the app itself.

If we remove the batch size knob, then this wouldn't be an issue, but I don't know if we need to worry enough to remove the knob.

Anyways, that's my current context, of course happy to find an even better way :) This seemed like the only way to materialize these properties of the BSP though.

Reply:
If you want a really quick fix then just block waiting for the CompleteableResultCode to finish. Nothing else is required. 

Blocking is a poor outcome here that can be easily avoided as per my other PR.

Reply:
Ah because we only have a single thread, if we're blocking then being unavailable isn't possible. Removed that check